### DISCUSSION TITLE
Is there a possibility of qutebrowser supporting keyboard interaction with DOM-less OpenGL-powered websites?

### DISCUSSION THREAD

**Discussion by interdisziplina** [qutebrowser/qutebrowser#7446] [General] [answered]
I have recently seen the [`egui`](https://lib.rs/crates/egui) framework for Rust that allows to build websites and native apps using just Rust, with (close to) no JS served using WASM. In contrast to DOM-manipulating frameworks like [`yew`](https://lib.rs/crates/yew), `egui` (and other similar emerging libraries) just work using OpenGL (afaik, I don't really know how it works underneath).

When I first saw the [demo page](https://egui.rs) on my mobile device, I was fascinated by how good it looked without any DOM and how interactive it was. However it is completely impossible to navigate this website in qutebrowser with the keyboard and needs to resort to use the mouse. Of course this makes sense, as qutebrowser tries to find links / actions via the DOM - and if there is none, no actions can be presented upon pressing `f` (or `:hint`, or `;h`, etc.).

I would like to know if there is a possibility of qutebrowser (and similar keyboard-centric browsers) improving this situation. I don't know if it is possible to extract information from the OpenGL hierarchy (if something like this even exists) or convert it to a DOM-like structure for this purpose. Or maybe there could be some sort of protocol egui and other libraries could implement that could communicate pressable elements on the page to qutebrowser (or any other interested recipients, for that matter)? 

I don't really know either whether this should be a discussion on qutebrowser, the egui framework or more general web standards stuff, but nonetheless it's important to think what the future of web development could be and how browsers must evolve to meet new standards and possibilities.

**Comment by The-Compiler** [comment, marked as answer]
> I would like to know if there is a possibility of qutebrowser (and similar keyboard-centric browsers) improving this situation. I don't know if it is possible to extract information from the OpenGL hierarchy (if something like this even exists) or convert it to a DOM-like structure for this purpose.

Not that I'm aware of. It simply paints things into a HTML canvas. The browser has no way (AFAIK) to know what's semantically being painted in there. I also don't really see how OpenGL comes into play.

> Or maybe there could be some sort of protocol egui and other libraries could implement that could communicate pressable elements on the page to qutebrowser (or any other interested recipients, for that matter)?

Well, that already exists, and it's called the DOM... I'd much rather see libraries like that use the browser to do drawing rather than reinventing the wheel. But if they insist on doing so, the most flexible way that comes to mind for it to tell a browser what elements are where, is indeed a (perhaps invisible/"fake") DOM.

> I don't really know either whether this should be a discussion on qutebrowser, the egui framework or more general web standards stuff, but nonetheless it's important to think what the future of web development could be and how browsers must evolve to meet new standards and possibilities.

It ties into much more than just qutebrowser. https://github.com/emilk/egui/issues/167 seems to have some related content as well.

I'm also not convinced this is the future of web development. It's a nice toy, but the overwhelming majority of websites won't have any benefits from it. For things that do (say, something like Photopea or so perhaps?), I could live with them not being keyboard-accessible via hints and such. Those are typically web applications (really using the browser as an alternative to a desktop GUI toolkit), and as such, will have their own shortcuts and such anyways.

**Reply by interdisziplina** [reply]
> Not that I'm aware of. It simply paints things into a HTML canvas. The browser has no way (AFAIK) to know what's semantically being painted in there. I also don't really see how OpenGL comes into play.

I didn't really research the OpenGL part when writing this and just saw it in a video. Turns out they just meant that WebGL, the tool for canvas painting, is based on OpenGL and can, if properly optimized from the WASM side, directly call it without any JavaScript interference - this works well with egui, as it is also built for desktop apps directly using OpenGL. An optimizing compiler could, in the case of egui, completely erase all calls to WebGL and just call OpenGL instead. Also AFAIK the canvas element is just an implementation detail (that may change in favor of a dedicated wasm canvas) and might change in the future (that's also why you can't read from the canvas at this point).

> Well, that already exists, and it's called the DOM... I'd much rather see libraries like that use the browser to do drawing rather than reinventing the wheel. But if they insist on doing so, the most flexible way that comes to mind for it to tell a browser what elements are where, is indeed a (perhaps invisible/"fake") DOM.

I generally agree that the DOM is probably a good idea. However I also think it is somewhat limiting and, most importantly, not really specified how elements are drawn in different browsers. WebGL / OpenGL are (again AFAIK) strong standards that won't behave differently just because you use Firefox instead of Chrome. I think web developers need more options (without reducing performance / usability) when it comes to web application design and specifically how elements are rendered on the page.  
A fake DOM might be a good idea, but I really don't know how this should be implemented on the egui side. Probably just a lot of invisible / background elements with some fixed x / y position? This would be an idea, but I don't know if it fits the immediate-mode rendering idea egui is going for. A least they allow for tabbing and some sort of accessibility features (as described in your linked issue). I'll probably file an issue at egui regarding this as a feature request.

> I'm also not convinced this is the future of web development. It's a nice toy, but the overwhelming majority of websites won't have any benefits from it. For things that do (say, something like Photopea or so perhaps?), I could live with them not being keyboard-accessible via hints and such. Those are typically web applications (really using the browser as an alternative to a desktop GUI toolkit), and as such, will have their own shortcuts and such anyways.

This is definitely true - I don't think a lot of websites will see a rewrite in egui or another library in this fashion. With that said, I really like the idea of WASM-only web apps just to reduce the monopoly JS has as the only frontend language (this makes it sound like it's some business, which it's at least not). WASM also has the benefit of not just being another layer on top of JS which is the majority of frameworks (Vue, React, Angular etc.) but instead a language-independant standard.  
Also, although many applications that may use this are based on desktop apps, not all of them will use keyboard shortcuts and some could definitely get better - the biggest example being forms (for which egui seems to be pretty optimized, considering all of its fancy input elements).

tl;dr: I really think you're right here that this is probably not that related to qutebrowser and that I should file an issue at egui to add an option for shadow DOM rendering. It probably is much more well-suited for alternatives to desktop applications just as Electron is supposed to web applications to the desktop. I still like the ideas behind egui and WASM-centered apps and think a combination of a DOM for accessibility and keyboard-centric purposes together with a freely interactable WebGL / OpenGL canvas is probably the best solution.

**Reply by The-Compiler** [reply]
> I think web developers need more options (without reducing performance / usability) when it comes to web application design and specifically how elements are rendered on the page.

I believe there's *a lot* you can do there via stylesheets. Yes, you still need to do some cross-browser testing, sure. But then you might just use a CSS framework abstracting this from you, instead of using a GUI framework doing its own pixel drawing. I really don't see the benefits of doing your own drawing for common UI elements. You'll need to reinvent a lot of things! Not just as far as qutebrowser is concerned, and also not just accessibility, but also things like having different keyboards for different inputs, dark mode, letting the user restyle things if they want, auto-fill, etc. etc. etc. These things are all solved by the browser in a standardized way, and those projects just throw all that away and reinvent the wheel. Might make sense if your goal is writing a browser game or something, but even for web applications I'm really not convinced.

> A fake DOM might be a good idea, but I really don't know how this should be implemented on the egui side. Probably just a lot of invisible / background elements with some fixed x / y position? This would be an idea, but I don't know if it fits the immediate-mode rendering idea egui is going for. A least they allow for tabbing and some sort of accessibility features (as described in your linked issue). I'll probably file an issue at egui regarding this as a feature request.

Sounds good to me! Yeah, the "fake DOM" idea is the first thing coming to mind. It would "just work" without any buy-in from the browser side. For anything needing browser-side support, it's unlikely that qutebrowser would be the first project supporting it. I'd rather focus our efforts elsewhere, than making a niche browser work better with a niche way of doing things in the web.

> to add an option for shadow DOM rendering

Note that [shadow DOM](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM) is already a pre-existing term for a web technology, so using that in this context will likely confuse people.
